source/common/http/conn_manager_impl.cc

```diff
@@ -1501,7 +1501,8 @@ void ConnectionManagerImpl::ActiveStream::addEncodedData(ActiveStreamEncoderFilt
                                                          Buffer::Instance& data, bool streaming) {
   if (state_.filter_call_state_ == 0 ||
       (state_.filter_call_state_ & FilterCallState::EncodeHeaders) ||
-      (state_.filter_call_state_ & FilterCallState::EncodeData)) {
+      (state_.filter_call_state_ & FilterCallState::EncodeData) ||
```

Comment:
The change here worth another PR as it at least seems risky and need a test case for here. Why did you need this?

Reply:
This is a tricky one.
When a gRPC server responds with 200, and then it sends a trailer with an error,
as in the `GrpcJsonTranscoderIntegrationTest.UnaryGetErrorInTrailerConvertedToJson` test,
or as it happens in this code:
```
func myHandler(w http.ResponseWriter, req *http.Request) {
  w.Header().Set("content-type", "application/grpc")
  w.Header().Set("trailer", "grpc-status")
  w.WriteHeader(200)
  // Then something goes wrong and we send a trailer with error:
  w.Header().Set("grpc-status", "5")
}
```
the transcoder is expected to send something like this:
```
$ echo -ne 'GET /func HTTP/1.1\r\nHost: localhost:8888\r\n\r\n' | nc localhost 8888
HTTP/1.1 404 Not Found
content-type: application/json
content-length: 10
date: Sat, 24 Aug 2019 20:56:19 GMT
x-envoy-upstream-service-time: 2
server: envoy

{"code":5}
```

However without this patch, first it sends the reply body with chunked transfer-encoding, and then it sends headers:
```
a
{"code":5}
HTTP/1.1 404 Not Found
content-type: application/json
content-length: 0
date: Sat, 24 Aug 2019 20:27:03 GMT
x-envoy-upstream-service-time: 12
server: envoy

```
(This "a" is the chunk size)

Supposedly when it does `encodeData()` in the `if (EncodeTrailers)` branch, it assumes that the headers were already sent, and maybe there was some part of reply-body sent, so it just dumps data as is.

Maybe there is a better option for fixing this, and yes I will also add a HttpConnectionManagerImplTest for this.

Reply:
I've removed the changes in HCM and disabled the feature in case when there is a separate trailer frame.

To sum it ip, an upstream has three ways to report an error:

1) Send a single header frame with end_stream flag which is also a trailer.
That's what happens when an unary function returns an error.

2) Send a header frame, send some data, and then send an error in a trailer.
We don't transcode the error into JSON because that would replace the data sent by upstream.

3) Send a header frame, and then send an error in a trailer.
This can happen when a gRPC sends metadata first, and then it encounters some error, e.g.
```
func (this *myServer) Func(...) (*pb.Ret, error) {
        ...
	grpc.SendHeader(ctx, someMetadata)
        ...
	return nil, someError
}
```
We don't support this yet, because HCM doesn't buffer data added from `encodeTrailers`.